src/audio/selector/selector_generic.c

```diff
@@ -84,16 +106,33 @@ static void sel_s32le_1ch(struct comp_dev *dev, struct audio_stream *sink,
 			  const struct audio_stream *source, uint32_t frames)
 {
 	struct comp_data *cd = comp_get_drvdata(dev);
-	int32_t *src;
-	int32_t *dest;
-	uint32_t i;
-	uint32_t j = 0;
-	uint32_t nch = cd->config.in_channels_count;
-
-	for (i = cd->config.sel_channel; i < frames * nch; i += nch) {
-		src = audio_stream_read_frag_s32(source, i);
-		dest = audio_stream_write_frag_s32(sink, j++);
-		*dest = *src;
+	int32_t *src = source->r_ptr;
+	int32_t *dest = sink->w_ptr;
+	int32_t *src_ch;
+	int nmax;
+	int i;
+	int n;
+	int processed = 0;
+	const int source_frame_bytes = audio_stream_frame_bytes(source);
+	const unsigned int nch = source->channels;
+	const unsigned int sel_channel = cd->config.sel_channel; /* 0 to nch - 1 */
+
+	assert(sel_channel < nch);
+	while (processed < frames) {
+		n = frames - processed;
+		nmax = audio_stream_bytes_without_wrap(source, src) / source_frame_bytes;
```

Comment:
since this will always be a multiple of 2 can we shift by sample_size (2 or 4) and mult by chans ? Means no divide.

Reply:
Samples count can be computed with shift but to get frames count would need to divide with channels. I'll see if I can build the inner loop other way to avoid divide here.

Reply:
I can't think of way to make this work with e.g. 3 channel to 1 without divide. The inner loop must operate with frames count and I can't change it to bytes (increments of frame bytes) because bytes max from source and sink are not "same thing". At least now I think so. The divide happens only max. twice per copy() so the cost is not high.